### PR TITLE
Feature-forward_onnx

### PR DESCRIPTION
**Pull request by VladislavPatrushev**

### CODE REVIEW

File: torchok/callbacks/model_checkpoint_with_onnx.py [hunk 1]
```diff
@@ -40,6 +41,10 @@ def _save_checkpoint(self, trainer: Trainer, filepath: str) -> None:
                 # DDP mode use some wrappers and we go down to BaseModel.
                 model = trainer.model.module.module if trainer.num_devices > 1 else trainer.model
                 input_tensors = model.input_tensors
+
+                if self.forward_onnx_outputs is not None:
```
Comment (PososikTeam) [inline torchok/callbacks/model_checkpoint_with_onnx.py:45]: Remove, need to get parameters from Task

File: torchok/callbacks/model_checkpoint_with_onnx.py [hunk 2]
```diff
@@ -40,6 +40,7 @@ def _save_checkpoint(self, trainer: Trainer, filepath: str) -> None:
                 # DDP mode use some wrappers and we go down to BaseModel.
                 model = trainer.model.module.module if trainer.num_devices > 1 else trainer.model
                 input_tensors = model.input_tensors
+
```
Comment (PososikTeam) [inline torchok/callbacks/model_checkpoint_with_onnx.py:43]: remove space line

File: torchok/tasks/classification.py [hunk 3]
```diff
@@ -46,16 +47,36 @@ def __init__(self, hparams: DictConfig):
         head_in_channels = self.pooling.out_channels
         self.head = HEADS.get(head_name)(in_channels=head_in_channels, **head_params)
 
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # check forward outputs is allowed or no.
+        self.forward_allowed_outputs = {'embeddings', 'predictions'}
+        self.forward_outputs = set(self._hparams.task.params.get('forward_outputs', 'predictions'))
+
+        if len(self.forward_allowed_outputs.intersection(self.forward_outputs)) == 0:
+            raise ValueError(f'{self.forward_outputs} is not allowed. '
```
Comment (PososikTeam) [inline torchok/tasks/classification.py:55]: `forward_outputs = {self.forward_outputs} is not allowed.`

File: torchok/tasks/classification.py [hunk 4]
```diff
@@ -46,16 +47,36 @@ def __init__(self, hparams: DictConfig):
...
+                             f'Available names: {self.forward_allowed_outputs}')
+
+    def forward(self, x: Tensor) -> Tensor:
```
Comment (PososikTeam) [inline torchok/tasks/classification.py:58]: change output type, to `Union[Tensor, Tuple[Tensor, ...]]`

File: torchok/tasks/segmentation.py [hunk 5]
```diff
@@ -43,12 +44,24 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
         """Forward with ground truth labels."""
-        input_data = batch['image']
-        target = batch['target']
+        input_data = batch.get('image')
+        target = batch.get('target')
         freeze_backbone = self._hparams.task.params.get('freeze_backbone', False)
         with torch.set_grad_enabled(not freeze_backbone and self.training):
             features = self.backbone.forward_features(input_data)
         neck_out = self.neck(features)
         prediction = self.head(neck_out)
```
Comment (PososikTeam) [inline torchok/tasks/segmentation.py:53]: Where the target in head? For example ArcFaceHead it need targets in training

File: torchok/callbacks/model_checkpoint_with_onnx.py [hunk 6]
```diff
@@ -39,8 +41,9 @@ def _save_checkpoint(self, trainer: Trainer, filepath: str) -> None:
             if self.export_to_onnx:
                 # DDP mode use some wrappers and we go down to BaseModel.
                 model = trainer.model.module.module if trainer.num_devices > 1 else trainer.model
-                input_tensors = model.input_tensors
-                model.to_onnx(filepath + self.ONNX_EXTENSION, (*input_tensors,), **self.onnx_params)
+                input_tensors = [getattr(model, name) for name in model.input_tensor_names]
+                model = model.as_module()[:-1] if self.remove_head else model.as_module()
+                torch.onnx.export(model, (*input_tensors,), filepath + self.ONNX_EXTENSION, **self.onnx_params)
```
Comment (Animatory) [inline torchok/callbacks/model_checkpoint_with_onnx.py:46]: Вместо `(*input_tensors,)` `tuple(input_tensors)`

File: torchok/callbacks/model_checkpoint_with_onnx.py [hunk 7]
```diff
@@ -39,8 +41,9 @@ def _save_checkpoint(self, trainer: Trainer, filepath: str) -> None:
             if self.export_to_onnx:
                 # DDP mode use some wrappers and we go down to BaseModel.
```
Comment (Animatory) [inline torchok/callbacks/model_checkpoint_with_onnx.py:42]: Add `,` after `wrappers`

File: torchok/tasks/classification.py [hunk 8]
```diff
@@ -55,14 +57,30 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
 
-    def forward_with_gt(self, batch: Dict[str, Union[torch.Tensor, int]]) -> Dict[str, torch.Tensor]:
+    def forward_with_gt(self, batch: Dict[str, Union[Tensor, int]]) -> Dict[str, Tensor]:
         """Forward with ground truth labels."""
-        input_data = batch['image']
-        target = batch['target']
-        features = self.backbone(input_data)
+        input_data = batch.get('image')
+        target = batch.get('target')
+        freeze_backbone = self._hparams.task.params.get('freeze_backbone', False)
```
Comment (Animatory) [inline torchok/tasks/classification.py:64]: Remove freeze backbones

File: torchok/tasks/classification.py [hunk 9]
```diff
@@ -55,14 +57,30 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        output = {'embeddings': embeddings, 'prediction': prediction}
+
+        if target is not None:
+            output['target'] = target
+
         return output
+
+    def as_module(self) -> nn.Sequential:
+        """Method for model representation as sequential of modules(need for checkpointing)."""
+        modules_for_checkpoint = []
```
Comment (Animatory) [inline torchok/tasks/classification.py:78]: Зачем тут лист?

File: torchok/tasks/segmentation.py [hunk 10]
```diff
@@ -43,12 +44,22 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        output = {'prediction': prediction}
+
+        if target is not None:
+            output['target'] = target
+
         return output
+
+    def as_module(self) -> nn.Sequential:
+        """Method for model representation as sequential of modules(need for checkpointing)."""
+        modules_for_checkpoint = []
```
Comment (Animatory) [inline torchok/tasks/segmentation.py:61]: Remove list

**Pull request merged**